overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,276 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapshotstate
+
+import (
+	"encoding/json"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/snapshotstate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var (
+	osRemove             = os.Remove
+	snapstateCurrentInfo = snapstate.CurrentInfo
+	configGetSnapConfig  = config.GetSnapConfig
+	configSetSnapConfig  = config.SetSnapConfig
+	backendOpen          = backend.Open
+	backendSave          = backend.Save
+	backendRestore       = (*backend.Reader).Restore // TODO: look into using an interface instead
+	backendCheck         = (*backend.Reader).Check
+	backendRevert        = (*backend.RestoreState).Revert // ditto
+	backendCleanup       = (*backend.RestoreState).Cleanup
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct {
+	runner *state.TaskRunner
+}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State) *SnapshotManager {
+	runner := state.NewTaskRunner(st)
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+	return &SnapshotManager{runner: runner}
+}
+
+// KnownTaskKinds is part of the overlord.StateManager interface.
+func (m *SnapshotManager) KnownTaskKinds() []string {
+	return m.runner.KnownTaskKinds()
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Stop() {
+	m.runner.Stop()
+}
+
+type snapshotState struct {
+	SetID    uint64   `json:"set-id"`
+	Snap     string   `json:"snap"`
+	Users    []string `json:"users,omitempty"`
+	Filename string   `json:"filename,omitempty"`
+}
+
+func filename(setID uint64, si *snap.Info) string {
```

Comment:
in the end I'm only using this function once (I'm pretty sure I used it more in the past, but this code has mutated over time). Probably easiest to just drop the function entirely.

Reply:
actually looking at the change, it would be marginally cleaner, but a lot harder to test in isolation (i'd have to export a bunch of stuff that I'd rather keep private)